### PR TITLE
Add test for successful vmcs_intel_x64::launch

### PR DESCRIPTION
**Pull request by connojd**
The test function vmcs_ut::test_launch_success mocks the
successful code path through vmcs_intel_x64::launch.

### CODE REVIEW

File: bfvmm/src/vmcs/test/test.h [hunk 1]
```diff
@@ -39,6 +39,10 @@ class vmcs_ut : public unittest
 
 private:
 
+    ///////////////////////////////////////////////////////////////////////////
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test.h]: get rid of the comments. group all of the tests by file. See here for example:
https://github.com/Bareflank/hypervisor/blob/master/bfvmm/src/memory_manager/test/test.h

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 2]
```diff
@@ -4,6 +4,7 @@
 // Copyright (C) 2015 Assured Information Security, Inc.
 // Author: Rian Quinn        <quinnr@ainfosec.com>
 // Author: Brendan Kerrigan  <kerriganb@ainfosec.com>
+// Author: Connor Davis <davisc@ainfosec.com>
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: 1) Alignment
2) Woot!!!

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 3]
```diff
@@ -20,3 +21,56 @@
...
+
+static void
+setup_vmcs_intrinsics(MockRepository &mocks, memory_manager *mm, intrinsics_intel_x64 *in)
+{
+    // Emulate the memory manager
+    mocks.OnCallFunc(memory_manager::instance).Return(mm);
+    mocks.OnCallOverload(mm, (uintptr_t(memory_manager::*)(void *))&memory_manager::virt_to_phys).Do(virt_to_phys_ptr);
+
+    // Setup MSR and vmread returns to mock a successful vmcs_intel_x64::filter_unsupported
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_BASIC_MSR).Return(0x7fffFFFF);
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: read / write MSR should call a read_msr / write_msr function with a switch statement that can pull from a std::map which you can prime with bad data if you need to. See here for example:

https://github.com/Bareflank/hypervisor/blob/master/bfvmm/src/serial/test/test_serial_port_intel_x64.cpp

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 4]
```diff
@@ -20,3 +21,56 @@
...
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_PINBASED_CTLS_MSR).Return(0x55555555aaaaAAAA);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_PROCBASED_CTLS_MSR).Return(0x55555555aaaaAAAA);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_PROCBASED_CTLS2_MSR).Return(0x55555555aaaaAAAA);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_EXIT_CTLS_MSR).Return(0x55555555aaaaAAAA);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_ENTRY_CTLS_MSR).Return(0x55555555aaaaAAAA);
+    mocks.OnCall(in, intrinsics_intel_x64::vmread).Return(0x5555555555555555);
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: same thing for vmread / vmwrite

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 5]
```diff
@@ -20,3 +21,56 @@
...
+    auto mm = mocks.Mock<memory_manager>();
+    auto in = bfn::mock_shared<intrinsics_intel_x64>(mocks);
+
+    setup_vmcs_intrinsics(mocks, mm, in.get());
+
+    RUN_UNITTEST_WITH_MOCKS(mocks, [&]
+    {
+        vmcs_intel_x64 vmcs(in);
+
+        auto host_state = std::make_shared<vmcs_intel_x64_state>();
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: These need to be mocked too

**Comment by rianquinn** [comment]
Looks good so far, minus the comments.

**Comment by connojd** [comment]
Changes from the above comments have been made

**Comment by connojd** [comment]
That's odd..all checks succeeded on my fork.

**Comment by brendank310** [comment]
@connojd I'm gonna have travis re-run the test on the PR, looks like it may have been a transient issue.

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 6]
```diff
@@ -20,3 +21,183 @@
...
+    mocks.OnCall(state_in, vmcs_intel_x64_state::cs_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ss_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ds_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ds_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::fs_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::gs_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ldtr_limit).Do(get_32bit_state);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::tr_limit).Do(get_32bit_state);
+
+    mocks.OnCall(state_in, vmcs_intel_x64_state::es_access_rights).Do(get_32bit_state);
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: each one of these needs it's own function which will be used later on in tests to be added in the future.

**Comment by connojd** [comment]
Thats the same error as the previous commit but with a different compiler

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 7]
```diff
@@ -20,3 +21,220 @@
...
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_pat_msr).Do(ia32_pat_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_efer_msr).Do(ia32_efer_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_perf_global_ctrl_msr).Do(ia32_perf_global_ctrl_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_cs_msr).Do(ia32_sysenter_cs_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_esp_msr).Do(ia32_sysenter_esp_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_eip_msr).Do(ia32_sysenter_eip_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_fs_base_msr).Do(ia32_fs_base_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_gs_base_msr).Do(ia32_gs_base_msr);
+
+    mocks.OnCall(state_in, vmcs_intel_x64_state::dump).Do(dump);
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: In a later update (not this PR), I would write this as:

```
mocks.OnCall(state_in, vmcs_intel_x64_state::dump);
```

There really is no needed to emulate this function (it's for debugging). Instead, if by registering OnCall with no `.XXX()`, you are telling Hippomocks to do nothing.

File: bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp [hunk 8]
```diff
@@ -20,3 +21,220 @@
...
+static void
+setup_vmcs_intrinsics(MockRepository &mocks, memory_manager *mm, intrinsics_intel_x64 *in)
+{
+    // Emulate the memory manager
+    mocks.OnCallFunc(memory_manager::instance).Return(mm);
+    mocks.OnCallOverload(mm, (uintptr_t(memory_manager::*)(void *))&memory_manager::virt_to_phys).Do(virt_to_phys_ptr);
+
+    // Setup MSR and vmread returns to mock a successful vmcs_intel_x64::filter_unsupported
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_BASIC_MSR).Return(0x7fffFFFF);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_PINBASED_CTLS_MSR).Return(0x55555555aaaaAAAA);
```
Comment (rianquinn) [inline bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp]: Still missing read_msr / write_msr functions for these. You can drop the `With()`, implement a read_msr / write_msr with `Do()` function and return the value you need for that specific test. The "setup" should provide a successful test case, and then each test only modifies one thing (i.e. the error your testing for).

**Comment by rianquinn** [comment]
Looking really good. Just a couple of minor things to cleanup and I will merge.

**Comment by connojd** [comment]
Should be good to go now

**Comment by rianquinn** [comment]
Ditch the `static void dump() {}` as stated in the previous comment and this L<great>TM. I have a PR in that adds Clang Tidy support, and addresses an issue with the virt -> phys conversion code. Once that is done (hopefully the push I just did passes this time), you should be good to rebase and update based on the new approach. If all goes well, we should be able to merge this PR today after you have rebased.

**Comment by rianquinn** [comment]
Now that the PR is in, I kicked off a new build of this PR which will likely fail as there will need to be some changes to support the new virt -> phys API, and probably some issues that Clang Tidy will complain about. Once the build is working again I will merge.

**Comment by connojd** [comment]
@rianquinn we've got a successful build after the rebase

**Pull request merged**